controllers/tenant_controller.go

```diff
@@ -0,0 +1,501 @@
+	logger.Info("starting finalization")
+	nss := &corev1.NamespaceList{}
+	if err := r.client.List(ctx, nss, client.MatchingFields{constants.RootNamespaces: tenant.Name}); err != nil {
+		return fmt.Errorf("failed to list namespaces: %w", err)
+	}
+	for _, ns := range nss.Items {
+		err := r.disownNamespace(ctx, &ns)
+		if err != nil {
+			return err
+		}
+		err = r.removeRBAC(ctx, tenant, &ns)
+		if err != nil {
+			return err
+		}
+	}
+	err := r.removeAppProject(ctx, tenant)
+	if err != nil {
+		return err
+	}
+
+	controllerutil.RemoveFinalizer(tenant, constants.Finalizer)
+	err = r.client.Update(ctx, tenant)
+	if err != nil {
+		logger.Error(err, "failed to remove finalizer")
+		return err
+	}
+	logger.Info("finished finalization")
+	return nil
+}
+
+func (r *TenantReconciler) patchNamespace(ctx context.Context, ns *accorev1.NamespaceApplyConfiguration) error {
+	obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(ns)
+	if err != nil {
+		return err
+	}
+	patch := &unstructured.Unstructured{
+		Object: obj,
+	}
+
+	var orig corev1.Namespace
+	err = r.client.Get(ctx, client.ObjectKey{Name: *ns.Name}, &orig)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	managed, err := accorev1.ExtractNamespace(&orig, constants.TenantFieldManager)
+	if err != nil {
+		return err
+	}
+
+	if equality.Semantic.DeepEqual(ns, managed) {
+		return nil
+	}
+
+	return r.client.Patch(ctx, patch, client.Apply, &client.PatchOptions{
+		FieldManager: constants.TenantFieldManager,
+		Force:        pointer.Bool(true),
+	})
+}
+
+func (r *TenantReconciler) patchRoleBinding(ctx context.Context, rb *acrbacv1.RoleBindingApplyConfiguration) error {
+	obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(rb)
+	if err != nil {
+		return err
+	}
+	patch := &unstructured.Unstructured{
+		Object: obj,
+	}
+
+	var orig rbacv1.RoleBinding
+	err = r.client.Get(ctx, client.ObjectKey{Namespace: *rb.Namespace, Name: *rb.Name}, &orig)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	managed, err := acrbacv1.ExtractRoleBinding(&orig, constants.TenantFieldManager)
+	if err != nil {
+		return err
+	}
+
+	if equality.Semantic.DeepEqual(rb, managed) {
+		return nil
+	}
+
+	return r.client.Patch(ctx, patch, client.Apply, &client.PatchOptions{
+		FieldManager: constants.TenantFieldManager,
+		Force:        pointer.Bool(true),
+	})
+}
+
+func rolesMap(delegates []cattagev1beta1.Delegate) map[string][]string {
+	result := make(map[string][]string)
+
+	for _, d := range delegates {
+		for _, role := range d.Roles {
+			result[role] = append(result[role], d.Name)
+		}
+	}
+	return result
+}
+
+func (r *TenantReconciler) reconcileNamespaces(ctx context.Context, tenant *cattagev1beta1.Tenant) error {
+	for _, ns := range tenant.Spec.RootNamespaces {
+		namespace := accorev1.Namespace(ns.Name)
+		labels := make(map[string]string)
+		for k, v := range r.config.Namespace.CommonLabels {
+			labels[k] = v
+		}
+		for k, v := range ns.Labels {
```

Comment:
It is better to document the labels and annotations of `RootNamespaces` supersedes `CommonLabels` and `CommonAnnotations`.